lib/livebook_web/plugs/proxy_plug.ex

```diff
@@ -0,0 +1,60 @@
+defmodule LivebookWeb.ProxyPlug do
+  @behaviour Plug
+  import Plug.Conn
+
+  alias Livebook.{App, Apps, Session, Sessions}
+  alias LivebookWeb.NotFoundError
+
+  @impl true
+  def init(opts), do: opts
+
+  @impl true
+  def call(%{path_info: ["sessions", id, "proxy" | path_info]} = conn, _opts) do
+    session = fetch_session!(id)
+    pid = fetch_proxy_handler!(session)
+    conn = prepare_conn(conn, path_info, ["sessions", id, "proxy"])
+    {conn, _} = Kino.Proxy.serve(pid, conn)
+
+    halt(conn)
+  end
+
+  def call(%{path_info: ["apps", slug, "proxy" | path_info]} = conn, _opts) do
```

Comment:
For now, let's only support "/apps/:slug/:session_id/proxy" and then we add support for "/apps/:slug/proxy" with automatic boot in a later PR.